### ISSUE TITLE
How to feed mqtt messages from within a fastapi websocket ? (ie. proxying)

### ISSUE THREAD

**Issue by MoonCactus** [sabuhish/fastapi-mqtt#52]
Sorry to post an issue - as it might better be a support request.

TLDR : how can I achieve the equivalent of having `@mqttclient.on_subscribe()` **within** fastapi `@app.websocket()` ?

I fail to understand if/how I can "locally" subscribe to topics from within my async fastapi websocket interfaces.

May be fastapi-mqtt is just not meant for this purpose.

Context is
  * parallel instances of an application of mine publish their states via MQTT, eg. `engine/213/progress` for the instance id 213.
  * I have a fastapi backend and multiple workers
  * I can subscribe and get notified of `engine/#` at the "root" level of my fast api app with `@mqttclient.on_subscribe()`
  * on the other hand I can feed data live to client from fastapi websocket  `@app.websocket()` async interfaces

But I wish I subscribed and transmitted the only topics that are actually open as webockets (1 websocket = 1 subscribe, activated on demand when a client connects to fastapi websocket for a given instance, then closed when the websocket goes away)

Ie. how can my back-end proxy the MQTT state of `engine/An/progress` only on demand, only from within the `@app.websocket("/engine/{An}/wsstatus")`, and only when a web client is connected ?

Meanwhile I feel like I have to`@mq.on_subscribe()` to all topics at the "root" level of my app, then try to asynchronously notify the currently opened fastapi websockets if something new arrives for them. This seems ugly and I am not even sure I can do it efficiently with multiple workers (may be gunicorn's `--preload` would help / using redis or memcached is overkill)

Well, or I revert to a non-fastapi mqtt client instanciated in the `@app.websocket()` !

(PS: I don't want to use paho-mqtt native websockets, it would require me to expose the mqtt network and bypass my fastapi auth)

**Comment by azogue** [comment]
Hi @MoonCactus,

The `@mqttclient.on_subscribe()` callback will trigger just to _confirm_ the subscription is OK, but all received MQTT messages will pass through `@mqttclient.on_message()` 😜. It is there where you have to _dispatch_ the received msgs in all subscribed topics to the open web sockets that match.

**It is not a problem to make dynamic subscriptions and unsubscriptions**, if you want to avoid listening to "xxx/#", but I imagine you would have to manage those separately, to avoid multiple subscriptions to the same topic (like: when opening the ws, check if already subscribed to topic, and subscribe only if not; and when closing the connection, do the opposite)

Then, by using some kind of async queue, you could dispatch the new MQTT messages through the ws connections as they come (Take a look to https://github.com/smithk86/asyncio-multisubscriber-queue). 

I did something similar in the recent past (sorry, not open source), but using server-sent-events instead of ws, and without the need of dynamic subscriptions, using that library to manage the current connections. It may serve as inspiration for you (but I imagine you already solved your problem, right?)
